flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptivebatch/util/PointwiseVertexInputInfoComputer.java

```diff
@@ -0,0 +1,257 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.scheduler.adaptivebatch.util;
+
+import org.apache.flink.runtime.executiongraph.ExecutionVertexInputInfo;
+import org.apache.flink.runtime.executiongraph.IndexRange;
+import org.apache.flink.runtime.executiongraph.JobVertexInputInfo;
+import org.apache.flink.runtime.executiongraph.VertexInputInfoComputationUtils;
+import org.apache.flink.runtime.scheduler.adaptivebatch.BlockingInputInfo;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static org.apache.flink.runtime.scheduler.adaptivebatch.DefaultVertexParallelismAndInputInfosDecider.MAX_NUM_SUBPARTITIONS_PER_TASK_CONSUME;
+import static org.apache.flink.runtime.scheduler.adaptivebatch.util.VertexParallelismAndInputInfosDeciderUtils.adjustToClosestLegalParallelism;
+import static org.apache.flink.runtime.scheduler.adaptivebatch.util.VertexParallelismAndInputInfosDeciderUtils.checkAndGetSubpartitionNum;
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/** Helper class that computes VertexInputInfo for pointwise input. */
+public class PointwiseVertexInputInfoComputer {
+
+    private static final Logger LOG =
+            LoggerFactory.getLogger(PointwiseVertexInputInfoComputer.class);
+
+    private final long dataVolumePerTask;
+
+    public PointwiseVertexInputInfoComputer(long dataVolumePerTask) {
+        this.dataVolumePerTask = dataVolumePerTask;
+    }
+
+    /**
+     * Decide parallelism and input infos, which will make the data be evenly distributed to
+     * downstream subtasks for POINTWISE, such that different downstream subtasks consume roughly
+     * the same amount of data.
+     *
+     * @param inputInfo The information of consumed blocking results
+     * @param parallelism The parallelism of the job vertex
+     * @return the vertex input info
+     */
+    public JobVertexInputInfo compute(BlockingInputInfo inputInfo, Integer parallelism) {
+
+        Map<Integer, long[]> subpartitionBytesByPartitionIndex =
+                inputInfo.getSubpartitionBytesByPartitionIndex();
+        int numPartitions = inputInfo.getNumPartitions();
+        int numSubPartitions = checkAndGetSubpartitionNum(Collections.singletonList(inputInfo));
+        // compute the size of each subpartition
```

Comment:
Looks to me the comment is not accurate. It just converts a 2D array into a 1D array.